File: day16/day16.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int PATTERN[4] = {0,1,0,-1};

char* read_input() {
    FILE* fp = fopen("input", "r");

    fseek(fp, 0, SEEK_END);
    size_t fsize = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    char* data = malloc(fsize);
    fread(data, sizeof(char), fsize, fp);
    data[fsize-1] = '\0';
    fclose(fp);

    return data;
}

int* parse_input(char* input) {
    size_t len = strlen(input);

    int* nums = malloc(len * sizeof(int));
    for (int i=0; i<len; i++){
        sscanf(input++, "%1d", &(nums[i]));
    }

    return nums;
}

int* simulate(char* input, int n_phases) {
    int* buf;
    int* signal = parse_input(input);
    size_t signal_l = strlen(input);
    int* new_signal = malloc(signal_l * sizeof(int));

    for (int phase=0; phase<n_phases; phase++){
        /*printf("Phase %d\n", phase);*/

        for (int digit=0; digit<signal_l; digit++){
            int total = 0;
            for (int i=0; i<signal_l; i++){
                /* printf("%d*%d + ", signal[i], at(pattern, i)); */
                int eh = signal[i] * PATTERN[((i+1) / (digit + 1)) % 4];
                total += eh;
                /* printf("total += %d\n", eh); */
            }
            /* printf(" = %d\n", abs(total) % 10); */
            new_signal[digit] = abs(total) % 10;
        }

        buf = signal;
        signal = new_signal;
        new_signal = buf;
    }

    free(new_signal);
    return signal;
}

void sim2(int* signal, size_t signal_l, int* newsignal) {
    long res = 0;
    for (int i = 0; i < signal_l; i++) {
        res += signal[i];
    }

    newsignal[0] = res % 10;
    for (int i = 0; i < signal_l - 1; i++){
        res -= signal[i];
        newsignal[i + 1] = res % 10;
    }
}

void part1() {
    char* input = read_input();
    int* signal = simulate(input, 100);
    printf("Part 1: ");
    for(int i=0; i<7; i++) printf("%d", signal[i]);
    printf("\n");

    free(signal);
    free(input);
}

void part2() {
    char* input = read_input();
    char offset_str[7];
    strncpy(offset_str, input, 7);
    int offset = atoi(offset_str);

    int copies = 5000;
    size_t input_len = strlen(input);
    int* signal = malloc(input_len * copies * sizeof(int));
    int* nums = parse_input(input);
    for (int i = 0; i<copies; i++) {
        memcpy(&signal[input_len * i], nums, input_len * sizeof(int));
    }
    offset -= input_len * 5000;
    /*printf("Offset: %d\n", offset);*/

    int* buf;
    int* newsignal = malloc(input_len * copies * sizeof(int));
    for (int phase = 0; phase < 100; phase++) {
        sim2(signal, input_len * 5000, newsignal);
        buf = signal;
        signal = newsignal;
        newsignal = buf;
    }
    free(newsignal);

    printf("Part 2: ");
    for(int i=offset; i<offset + 8; i++) printf("%d", signal[i]);
    printf("\n");

    free(nums);
    free(input);
    free(signal);
}

int main(int argc, char** argv) {
    part1();
    part2();

    return 0;
}

File: day16/day16.py
import math
import itertools
from functools import cache


PATTERN = [0, 1, 0, -1]


def repeat(times):
    for count in itertools.count(1):
        idx = (count // times) % len(PATTERN)
        yield PATTERN[idx]


def calc(signal, idx):
    return abs(sum(a * b for a, b in zip(signal, repeat(idx)))) % 10


def simphase(signal):
    return tuple(calc(signal, idx) for idx, digit in enumerate(signal, start=1))


def part1(signal, phases=100):
    signal = tuple(map(int, signal))

    for phase in range(phases):
        # print(phase)
        signal = simphase(signal)

    return "".join(map(str, signal[:7]))


def sim2(signal):
    res = sum(signal)
    new = [res % 10]
    for t in signal[:-1]:
        res -= t
        new.append(res % 10)

    return new


def part2(signal, phases=100):
    offset = int(signal[:7])
    signal = tuple(map(int, signal)) * 5000
    breakpoint()
    offset -= len(signal)

    for phase in range(phases):
        signal = sim2(signal)

    # print("out", "".join(map(str, signal)))
    return "".join(map(str, signal[offset : offset + 8]))


def read_input():
    with open("input") as f:
        return f.read().strip()


if __name__ == "__main__":
    print("Part 1:", part1(read_input(), 100))
    print("Part 2:", part2(read_input(), 100))
